source/common/common/perf_annotation.cc

```diff
@@ -0,0 +1,130 @@
+  std::unique_lock<std::mutex> lock(context->mutex_);
+#endif
+
+  // The map is from category/description -> [duration, time]. Reverse-sort by duration.
+  std::vector<const DurationCountMap::value_type*> sorted_values;
+  sorted_values.reserve(context->duration_count_map_.size());
+  for (const auto& iter : context->duration_count_map_) {
+    sorted_values.push_back(&iter);
+  }
+  std::sort(
+      sorted_values.begin(), sorted_values.end(),
+      [](const DurationCountMap::value_type* a, const DurationCountMap::value_type* b) -> bool {
+        return a->second.first > b->second.first;
+      });
+
+  // Organize the report so it lines up in columns. Note that the widest duration comes first,
+  // though that may not be descending order of calls or per_call time, so we need two passes
+  // to compute column widths. First collect the column headers and their widths.
+  static const char* headers[] = {"Duration(us)", "# Calls", "per_call(ns)",
+                                  "Category / Description"};
+  constexpr int num_columns = ARRAY_SIZE(headers);
+  size_t widths[num_columns];
+  std::vector<std::string> columns[num_columns];
+  for (size_t i = 0; i < num_columns; ++i) {
+    columns[i].push_back(headers[i]);
+    widths[i] = strlen(headers[i]);
+  }
+
+  // Compute all the column strings and their max widths.
+  for (const auto& p : sorted_values) {
+    const DurationCount& duration_count = p->second;
+    std::chrono::nanoseconds duration = duration_count.first;
+    uint64_t count = duration_count.second;
+    columns[0].push_back(
+        std::to_string(std::chrono::duration_cast<std::chrono::microseconds>(duration).count()));
+    columns[1].push_back(std::to_string(count));
+    columns[2].push_back(
+        (count == 0)
+            ? "NaN"
+            : std::to_string(
+                  std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count() / count));
+    columns[3].push_back(p->first);
+    for (size_t i = 0; i < num_columns; ++i) {
+      widths[i] = std::max(widths[i], columns[i].back().size());
+    }
+  }
+
+  // Write out the table.
+  for (size_t row = 0; row < columns[0].size(); ++row) {
+    for (size_t i = 0; i < num_columns; ++i) {
+      const std::string& str = columns[i][row];
+      // Right-justify all but last column by appending the number of spaces needed to bring
+      // it inline with the largest.
+      if (i != (num_columns - 1)) {
+        out.append(widths[i] - str.size(), ' ');
+        absl::StrAppend(&out, str, "  ");
+      } else {
+        absl::StrAppend(&out, str, "\n");
+      }
+    }
+  }
+  return out;
+}
+
+void PerfAnnotationContext::clear() {
+  PerfAnnotationContext* context = getOrCreate();
+#if PERF_THREAD_SAFE
+  std::unique_lock<std::mutex> lock(context->mutex_);
+#endif
+  context->duration_count_map_.clear();
+}
+
+PerfAnnotationContext* PerfAnnotationContext::getOrCreate() {
+  static PerfAnnotationContext* context = new PerfAnnotationContext();
```

Comment:
Is there any reason we can't just subclass the [ThreadSafeSingleton](https://github.com/envoyproxy/envoy/blob/master/source/common/singleton/threadsafe_singleton.h#L24) to keep this singleton implementation consistent with others like it across Envoy?